web/federate.go

```diff
@@ -179,7 +192,7 @@ func (h *Handler) federation(w http.ResponseWriter, req *http.Request) {
 			}
 			if l.Name == labels.MetricName {
 				nameSeen = true
-				if l.Value == lastMetricName {
+				if l.Value == lastMetricName && lastWasHistogram == isHistogram {
```

Comment:
Another problem here is that we don't detect if there are gauge histograms and counter histograms under the same name. Or in other words: In the unlikely case that, under the same name, we change from gauge histogram to counter histogram, or vice versa, we have to do the same thing as if we change from conventional metric to histogram, namely starting a new metric family.

Reply:
And somewhere here, we need a code comment saying that this will create multiple metric families of the same name, which would be invalid in a "proper" exposition, but Prometheus ingests it just fine, and therefore we bend the rules for federation here.

Reply:
And final note: If I read the code correctly, we could even run into the weird case that we flip flop between different types multiple times, and then we create a new metric family each time. But even that should be fine. Just to confirm that we have the same understanding.

Reply:
> Just to confirm that we have the same understanding.

Yes

Reply:
It now looks for changes in histogram type as well. We cannot test it at the moment because PromQL only returns an unknown counter reset.